kpi/deployment_backends/base_backend.py

```diff
@@ -90,34 +103,97 @@ def get_data(self,
 
         return value
 
-    def get_submission(self, pk, requesting_user_id,
-                       format_type=INSTANCE_FORMAT_TYPE_JSON, **kwargs):
+    @abc.abstractmethod
+    def delete_submission(self, submission_id: int, user: 'auth.User') -> dict:
+        pass
+
+    @abc.abstractmethod
+    def duplicate_submission(
+        self,  submission_id: int, user: 'auth.User', **kwargs
+    ) -> dict:
+        pass
+
+    @abc.abstractmethod
+    def get_data_download_links(self):
+        pass
+
+    @abc.abstractmethod
+    def get_enketo_submission_url(
+        self, submission_id: int, user: 'auth.User', params: dict = None
+    ) -> dict:
         """
-        Returns submission if `pk` exists otherwise `None`
-
-        Args:
-            pk (int): Submission's primary key
-            requesting_user_id (int)
-            format_type (str): INSTANCE_FORMAT_TYPE_JSON|INSTANCE_FORMAT_TYPE_XML
-            kwargs (dict): Filters to pass to MongoDB. See
-                https://docs.mongodb.com/manual/reference/operator/query/
-
-        Returns:
-            (dict|str|`None`): Depending on `format_type`, it can return:
-                - Mongo JSON representation as a dict
-                - Instance's XML as string
-                - `None` if doesn't exist
+        Return a formatted dict to be passed to a Response object
         """
+        pass
 
-        submissions = list(self.get_submissions(requesting_user_id,
-                                                format_type, [int(pk)],
-                                                **kwargs))
+    @abc.abstractmethod
+    def get_enketo_survey_links(self):
+        pass
+
+    def get_submission(self,
+                       submission_id: int,
+                       user: 'auth.User',
+                       format_type: str = INSTANCE_FORMAT_TYPE_JSON,
+                       **kwargs: dict) -> Union[dict, str, None]:
+        """
+        Returns the corresponding submission whose id equals `pk` and which
+        `user` is allowed to access.
+        Otherwise, it returns `None`.
+        The format `format_type` can be either:
+        - 'json' (See `kpi.constants.INSTANCE_FORMAT_TYPE_JSON)
+        - 'xml' (See `kpi.constants.INSTANCE_FORMAT_TYPE_XML)
+
+        MongoDB filters can be passed through `kwargs` to narrow down the
+        result.
+        """
+
+        submissions = list(
+            self.get_submissions(
+                user, format_type, [int(submission_id)], **kwargs
+            )
+        )
         try:
             return submissions[0]
         except IndexError:
             pass
         return None
 
+    @abc.abstractmethod
+    def get_submission_detail_url(self, submission_id: int) -> str:
+        pass
+
+    def get_submission_validation_status_url(self, submission_id: int) -> str:
+        # This does not really need to be implemented.
```

Comment:
Comment has been removed. This method was in MockBackend at the first place. I moved it to `base_backend` since it shares exactly the same code between deployment back-end classes. I should have removed the comment as it's totally useless now.